substrate/client/network/src/litep2p/discovery.rs

```diff
@@ -715,3 +717,166 @@ impl Stream for Discovery {
+					listen_addresses.write().insert(address.clone());
+				});
+
+				// Except the first peer, all other peers know the first peer addresses.
+				if i == 0 {
+					log::info!(target: LOG_TARGET, "First peer is {peer_id:?} with addresses {addresses:?}");
+					known_peers.insert(peer_id, addresses.clone());
+				} else {
+					let (peer, addresses) = known_peers.iter().next().unwrap();
+
+					let result = litep2p.add_known_address(*peer, addresses.into_iter().cloned());
+
+					log::info!(target: LOG_TARGET, "{peer_id:?}: Adding known peer {peer:?} with addresses {addresses:?} result={result:?}");
+
+				}
+
+				(peer_id, litep2p, discovery)
+			})
+			.collect::<Vec<_>>();
+
+		let total_peers = backends.len() as u32;
+		let remaining_peers =
+			backends.iter().map(|(peer_id, _, _)| *peer_id).collect::<HashSet<_>>();
+
+		let first_peer = *known_peers.iter().next().unwrap().0;
+
+		// Each backend must discover the whole network.
+		let mut futures = FuturesUnordered::new();
+		let num_finished = Arc::new(AtomicU32::new(0));
+
+		for (peer_id, mut litep2p, mut discovery) in backends {
+			// Remove the local peer id from the set.
+			let mut remaining_peers = remaining_peers.clone();
+			remaining_peers.remove(&peer_id);
+
+			let num_finished = num_finished.clone();
+
+			let future = async move {
+				log::info!(target: LOG_TARGET, "{peer_id:?} starting loop");
+
+				if peer_id != first_peer {
+					log::info!(target: LOG_TARGET, "{peer_id:?} dialing {first_peer:?}");
+					litep2p.dial(&first_peer).await.unwrap();
+				}
+
+				loop {
+					// We need to keep the network alive until all peers are discovered.
+					if num_finished.load(std::sync::atomic::Ordering::Relaxed) == total_peers {
+						log::info!(target: LOG_TARGET, "{peer_id:?} all peers discovered");
+						break
+					}
+
+					tokio::select! {
+						// Drive litep2p backend forward.
+						event = litep2p.next_event() => {
+							log::info!(target: LOG_TARGET, "{peer_id:?} Litep2p event: {event:?}");
+						},
+
+						// Detect discovery events.
+						event = discovery.next() => {
+							match event.unwrap() {
+								// We have discovered the peer via kademlia and established
+								// a connection on the identify protocol.
+								DiscoveryEvent::Identified { peer, .. } => {
+									log::info!(target: LOG_TARGET, "{peer_id:?} Peer {peer} identified");
+
+									remaining_peers.remove(&peer);
+
+									if remaining_peers.is_empty() {
+										log::info!(target: LOG_TARGET, "{peer_id:?} All peers discovered");
+
+										num_finished.fetch_add(1, std::sync::atomic::Ordering::AcqRel);
+									}
+								},
+
+								event => {
+									log::info!(target: LOG_TARGET, "{peer_id:?} Discovery event: {event:?}");
+								}
+							}
+						}
+					}
+				}
+			};
+
+			futures.push(future);
+		}
+
+		// Futures will exit when all peers are discovered.
+		tokio::time::timeout(Duration::from_secs(60), futures.next())
```

Comment:
Just curious, what is the expected time to finish?

Reply:
This roughly finishes in 10s on my machine, I've bumped this a bit as Im afraid it will become flaky on the CI